### ISSUE TITLE
HAProxy Graceful Shutdown

### ISSUE THREAD

**Issue by avmnu-sng** [jcmoraisjr/haproxy-ingress#1030]
Quoting the [HAProxy management doc](https://docs.haproxy.org/2.4/management.html#4):
> HAProxy supports a graceful and a hard stop. The hard stop is simple, when the
SIGTERM signal is sent to the haproxy process, it immediately quits and all
established connections are closed. The graceful stop is triggered when the
SIGUSR1 signal is sent to the haproxy process. It consists in only unbinding
from listening ports, but continue to process existing connections until they
close. Once the last connection is closed, the process leaves.

In the https://github.com/jcmoraisjr/haproxy-ingress/blob/master/rootfs/haproxy-shutdown.sh, why are we not sending **_SIGUSR1_**?

```bash
pid=$(cat "$HAPROXY_PID")

# Send SIGUSR1
kill -USR1 $pid

# Wait to terminate
while $(kill -0 $PID 2>/dev/null); do
  sleep 1
done

echo "PID: $pid"
```

**Comment by jcmoraisjr** [comment]
Hi, this is a script that shuts down haproxy when the controller is stopping, e.g., on a rolling update, after a possible wait for shutdown has timed out. There's another script that reloads haproxy, via -sf command-line option, which in turn uses SIGUSR1 to notificate the old process.

**Comment by avmnu-sng** [comment]
Hi, I am talking about shutting down scenario only. If we do `kill $pid`, HAProxy will drop the ongoing requests also.

**Comment by jcmoraisjr** [comment]
This is what we want due to the phase of the controller shutdown that this script is launched. I'm planning not to improve ha on this way, but instead following some ideas from #1028, which means, you need to drop incoming connections early from a fronting lb or dns, and after that start the controller and haproxy shutdown process. We're missing some ha related doc, and maybe a few more improvements in the code as well. I'll update #1028 as soon as we've news on the subject.

**Comment by avmnu-sng** [comment]
That's true. To properly handle graceful, we need to drop incoming traffic.

As a hack/workaround, I was thinking of running a `preStop` hook to block incoming traffic:
```bash
iptables -A INPUT -p tcp --dport 80 -j DROP
sleep t # allows ongoing requests to finish 
````
The problem is that if we don't have another controller to receive traffic, all the incoming requests fail. Probably:
```bash
sleep t1 # Let the new controller pod come up
iptables -A INPUT -p tcp --dport 80 -j DROP
sleep t2 # allows ongoing requests to finish 
```
@jcmoraisjr Thoughts?
